matchbox_signaling/src/signaling_server/builder.rs

```diff
@@ -0,0 +1,139 @@
+use crate::{
+    signaling_server::{
+        callbacks::{Callback, SharedCallbacks},
+        handlers::{ws_handler, WsUpgradeMeta},
+        NoOpCallouts, NoState,
+    },
+    topologies::{SignalingStateMachine, SignalingTopology},
+    SignalingCallbacks, SignalingServer, SignalingState,
+};
+use axum::{routing::get, Extension, Router};
+use std::{marker::PhantomData, net::SocketAddr};
+use tower_http::{
+    cors::{Any, CorsLayer},
+    trace::{DefaultOnResponse, TraceLayer},
+    LatencyUnit,
+};
+use tracing::Level;
+
+/// Builder for [`SignalingServer`]s.
+///
+/// Begin with [`SignalingServerBuilder::new`] and add parameters before calling
+/// [`SignalingServerBuilder::build`] to produce the desired [`SignalingServer`].
+pub struct SignalingServerBuilder<Topology, Cb = NoOpCallouts, S = NoState>
+where
+    Topology: SignalingTopology<Cb, S>,
+    Cb: SignalingCallbacks,
+    S: SignalingState,
+{
+    /// The socket address to broadcast on
+    pub(crate) socket_addr: SocketAddr,
+
+    /// The router used by the signaling server
+    pub(crate) router: Router,
+
+    /// Shared callouts used by all signaling servers
+    pub(crate) shared_callbacks: SharedCallbacks,
+
+    /// The callbacks used by the signaling server
+    pub(crate) callbacks: Cb,
+
+    /// The state machine that runs a websocket to completion, also where topology is implemented
+    pub(crate) topology: Topology,
+
+    /// Arbitrary state accompanying a server
+    pub(crate) state: PhantomData<S>,
+}
+
+impl<Topology, Cb, S> SignalingServerBuilder<Topology, Cb, S>
+where
+    Topology: SignalingTopology<Cb, S>,
+    Cb: SignalingCallbacks,
+    S: SignalingState,
+{
+    /// Creates a new builder for a [`SignalingServer`].
+    pub fn new(socket_addr: impl Into<SocketAddr>, topology: Topology, state: S) -> Self {
+        Self {
+            socket_addr: socket_addr.into(),
+            router: Router::new()
+                .route("/", get(ws_handler::<Cb, S>))
+                .route("/:path", get(ws_handler::<Cb, S>))
+                .with_state(state),
+            shared_callbacks: SharedCallbacks::default(),
+            callbacks: Cb::default(),
+            topology,
+            state: PhantomData,
+        }
+    }
+
+    /// Modify the router with a mutable closure. This is where one may apply middleware or other
+    /// layers to the Router.
+    pub fn middleware(mut self, mut alter: impl FnMut(&mut Router)) -> Self {
+        alter(&mut self.router);
+        self
+    }
+
+    /// Change the topology.
+    pub fn topology(mut self, topology: Topology) -> Self {
+        self.topology = topology;
+        self
+    }
```

Comment:
I might be missing something very obvious here, but how does this work when we have the type state pattern locking us to a given Topology?

Reply:
In my mind, originally this was useful. I don't know if it's still useful.

My original thought was, if you own your own topology, like OneToOne, you'd still use FullMeshState and FullMeshCallbacks likely. So changing the topology would make a Builder<OneToOne, FullMeshCallbacks, FullMeshState> if you started with a full_mesh_builder and changed the topology to OneToOne.

I do agree this seems useless in practice... it should be removed. Creating a new topology is an endeavor and something not likely someone would do, as well this would probably confuse them.

Reply:
> started with a full_mesh_builder and changed the topology to OneToOne

I don't believe this is possible as `topology` must be of type `Topology`, which is already bound to the current topology. This could be fixed by having the function create a new correctly typed builder, but I don't think it's really worth the effort when anyone creating a Topology can use an extension trait to create a `new_xyz` function on the builder